app/code/Magento/Developer/Model/TemplateEngine/Plugin/DebugHints.php

```diff
@@ -94,15 +122,30 @@ public function afterCreate(
         $storeCode = $this->storeManager->getStore()->getCode();
         if ($this->scopeConfig->getValue($this->debugHintsPath, ScopeInterface::SCOPE_STORE, $storeCode)
             && $this->devHelper->isDevAllowed()) {
-            $showBlockHints = $this->scopeConfig->getValue(
-                self::XML_PATH_DEBUG_TEMPLATE_HINTS_BLOCKS,
+            $debugHintsShowWithParameter = $this->scopeConfig->getValue(
+                $this->debugHintsShowWithParameter,
+                ScopeInterface::SCOPE_STORE,
+                $storeCode
+            );
+            $debugHintsParameter = $this->scopeConfig->getValue(
+                $this->debugHintsParameter,
                 ScopeInterface::SCOPE_STORE,
                 $storeCode
             );
-            return $this->debugHintsFactory->create([
-                'subject' => $invocationResult,
-                'showBlockHints' => $showBlockHints,
-            ]);
+            $debugHintsParameterInUrl = $this->http->getParam('templatehints');
+            if ((!$debugHintsShowWithParameter) ||
```

Comment:
This condition looks hard to read and understand, increasing possibility of an error in future refactoring. Please attempt to simplify it.